test/src/main/java/org/axonframework/test/matchers/DeepEqualsMatcher.java

```diff
@@ -0,0 +1,149 @@
+
+    private static final Logger logger = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+    private final T expected;
+    private final FieldFilter filter;
+
+    private boolean noneMatchingTypes = false;
+    private boolean noneMatchingEquals = false;
+
+    private Field failedField;
+    private Object failedFieldExpected;
+    private Object failedFieldActual;
+    private boolean failedForAccessibilityException;
+
+
+    /**
+     * Construct a {@link DeepEqualsMatcher} that will match an {@code actual} value with the given {@code expected}.
+     *
+     * @param expected The object to match with during {@link #matches(Object)}.
+     */
+    public DeepEqualsMatcher(T expected) {
+        this(expected, AllFieldsFilter.instance());
+    }
+
+    /**
+     * Construct a {@link DeepEqualsMatcher} that will match an {@code actual} value with the given {@code expected}.
+     *
+     * @param expected The object to match with during {@link #matches(Object)}.
+     * @param filter   The filter describing the fields to include or exclude in the comparison.
+     */
+    public DeepEqualsMatcher(T expected, FieldFilter filter) {
+        assertNonNull(expected, "The expected value should be non-null.");
+        this.expected = expected;
+        this.filter = filter;
+    }
+
+    @Override
+    public boolean matches(Object actual) {
+        if (!matchingTypes(actual)) {
+            noneMatchingTypes = true;
+            return false;
+        }
+        if (Objects.equals(expected, actual)) {
+            return true;
+        }
+        if (hasEqualsMethod(actual.getClass())) {
+            // Expected does not equal actual, and equals is implemented. Hence, we should not perform field equality.
+            noneMatchingEquals = true;
+            return false;
+        }
+        return matchingFields(expected.getClass(), expected, actual);
+    }
+
+    private boolean matchingTypes(Object actual) {
+        return expected.getClass().isInstance(actual) && expected.getClass().equals(actual.getClass());
+    }
+
+    private boolean matchingFields(Class<?> aClass, Object expectedValue, Object actual) {
+        boolean match = true;
+        for (Field field : aClass.getDeclaredFields()) {
+            if (filter.accept(field)) {
+                try {
+                    field.setAccessible(true);
+                    Object expectedFieldValue = field.get(expectedValue);
+                    Object actualFieldValue = field.get(actual);
+                    if (!Objects.deepEquals(expectedFieldValue, actualFieldValue)) {
+                        failedField = field;
+                        failedFieldExpected = expectedFieldValue;
+                        failedFieldActual = actualFieldValue;
+                        return false;
+                    }
+                } catch (Exception e) {
+                    if (e.getClass().getSimpleName().equals("InaccessibleObjectException")) {
```

Comment:
In general, it's good practice to make call `.equals()` on static parts, to avoid NPE. In this case it probably doesn't make much of a difference, though.
Recommended: `if ("InaccessibleObjectException".equals(...) )`

Reply:
You're completely right. Flipped the equals check.